lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+import logging
+from dataclasses import dataclass
+from itertools import chain
+from typing import Dict, List, Tuple, Union
+
+import networkx as nx
+
+from lineapy.api.api import get
+from lineapy.api.api_classes import LineaArtifact
+from lineapy.data.types import LineaID
+from lineapy.graph_reader.node_collection import NodeCollectionType
+from lineapy.graph_reader.session_artifacts import SessionArtifacts
+from lineapy.plugins.task import TaskGraphEdge
+from lineapy.plugins.utils import load_plugin_template
+from lineapy.utils.logging_config import configure_logging
+from lineapy.utils.utils import prettify
+
+logger = logging.getLogger(__name__)
+configure_logging()
+
+
+@dataclass
+class ArtifactCollection:
+    """
+    `ArtifactCollection` can be thought of as a box where the inserted group of artifacts and
+    their graph(s) get refactored into reusable components (i.e., functions with non-overlapping
+    operations). With this modularization, it can then support various downstream code generation
+    tasks such as pipeline file writing.
+
+    For now, `ArtifactCollection` is meant to be kept and used as an abstraction/tool for internal
+    dev use only. That is, the class and its methods will NOT be exposed directly to the user.
+    Instead, it is intended to be used by/in/for other user-facing APIs.
+    """
+
+    def __init__(self, artifacts=List[Union[str, Tuple[str, int]]]) -> None:
+        self.session_artifacts: Dict[LineaID, SessionArtifacts] = {}
+        self.art_name_to_node_id: Dict[str, LineaID] = {}
+        self.node_id_to_session_id: Dict[LineaID, LineaID] = {}
+
+        artifacts_by_session: Dict[LineaID, List[LineaArtifact]] = {}
+
+        # Retrieve artifact objects and group them by session ID
+        for art_entry in artifacts:
+            # Construct args for artifact retrieval
+            args = {}
+            if isinstance(art_entry, str):
+                args["artifact_name"] = art_entry
+            elif isinstance(art_entry, tuple):
+                args["artifact_name"] = art_entry[0]
+                args["version"] = art_entry[1]
+            else:
+                raise ValueError(
+                    "An artifact should be passed in as a string or (string, integer) tuple."
+                )
+
+            # Retrieve artifact
+            try:
+                # art = get(**args)
```

Comment:
nit remove this line

Reply:
@andycui97 Good catch!

@mingjerli Is there a reason why we want to replace `art = get(**args)` with the more explicit code (I thought we have full control over what's packed inside `args`)?

Reply:
If my memory is correct, mypy was complaining about this.